base/genericmemory.jl

```diff
@@ -335,3 +339,94 @@ end
     end
 end
 view(m::GenericMemory, inds::Colon) = view(m, eachindex(m))
+
+# modify, swap and replace at index
+function getindex_atomic(
+    mem::GenericMemory,
+    i::Int,
+    order=default_access_order(mem)
+)
+    memref = memoryref(mem, i)
+    return memoryrefget(memref, order, @_boundscheck)
+end
+
+function setindex_atomic!(
+    mem::GenericMemory,
+    i::Int,
+    val,
+    order=default_access_order(mem)
+)
```

Comment:
Since the index is already typically Vararg, I don't know that `order` can be optional to specify for the function (I think that is the advantage of the macro syntax is that it can supply the default):
```suggestion
setindex_atomic!(mem::GenericMemory, order::Symbol, val, inds...) = setindex_atomic!(mem, order, val, _to_linear_index(A, inds...))

function setindex_atomic!(mem::GenericMemory, order::Symbol, val, i::Int)
```
But instead of that specific former method there, you can specify that method in general following the pattern of isassigned in base/multidimensional.jl

Reply:
Can `set`/`getindex` be performed atomically with multiple indices, or rather each of the operations will be atomic, with their order unspecified? Then the name of the function would be a misnomer...

I like the suggested order of arguments (with `order` as the second one).

Reply:
the key point here is that for AtomicMemory, the only part that needs to be atomic is the memoryrefset (it replace etc) nine of the computation of which index needs to be set needs to be atomic because we're only creating atomicity of the elements

Reply:
@oscardssmith sure, but then `setindex_atomic!(at_mem, :sequentially_consistent, [a,b], [1,2])` is not atomic, i.e. it consists of two atomic operations which can be reordered according to the order provided. That's why I said that the name would be a misnomer.

Reply:
@vtjnash Looking at base/multidimensional.jl I'm still a bit confused.
Do you want me to replace e.g. the existing `getindex_atomic` by the following?
```julia
function getindex_atomic(mem::GenericMemory, order::Symbol, i::Union{Integer,CartesianIndex}...)
    return getindex_atomic(mem, order, CartesianIndex(to_indices(mem, i)))
end
function getindex_atomic(mem::GenericMemory, order::Symbol, i::CartesianIndex)
    return getindex_atomic(mem, order, Tuple(i)...)
end
function getindex_atomic(mem::GenericMemory, order::Symbol, i::Integer...)
    idcs = to_indices(mem, i)
    S = IndexStyle(mem)
    if isa(S, IndexLinear)
        return getindex_atomic(mem, order, _to_linear_index(mem, idcs...))
    else
        return getindex_atomic(mem, order, _to_subscript_indices(mem, idcs...)...)
    end
end
function getindex_atomic(mem::GenericMemory, order::Symbol, i::Int)
    memref = memoryref(mem, i)
    return memoryrefget(memref, order, @_boundscheck)
end
```
or something yet different? Since `IndexStyle of `GenericMemory` is defined to be `IndexLinear` by abstract method, the other branch will be empty until someone actually implements some funky `GenericMemory`.

Similar methods will need to be added for `setindex_atomic!`, `setindexonce_atomic!`, `modifyindex_atomic!`, `swapindex_atomic!`, and `replaceindex_atomic!`.

Is that what you meant?

Reply:
exactly

Reply:
3×6 methods of signature × function_name ;)

Reply:
So I wanted to do this in `genericmemory.jl`
```julia
## vararg methods generation
for (fn, args) in (
    :(getindex_atomic(mem::GenericMemory, order::Symbol, i::Int)),
    :(setindex_atomic!(mem::GenericMemory, order::Symbol, val, i::Int)),
    :(setindexonce_atomic!(mem::GenericMemory, success_order::Symbol, fail_order::Symbol, val, i::Int)),
    :(modifyindex_atomic!(mem::GenericMemory, order::Symbol, op, val, i::Int)),
    :(swapindex_atomic!(mem::GenericMemory, order::Symbol, val, i::Int)),
    :(replaceindex_atomic!(mem::GenericMemory, success_order::Symbol, fail_order::Symbol, expected, desired, i::Int,)),
)
    fn = ex.args[1]
    args = ex.args[2:end-1]

    @eval begin
        function $fn($(args...), i::Union{Integer,CartesianIndex}...)
            return $fn($(args...), CartesianIndex(to_indices($(args[1]), i)))
        end

        function $fn($(args...), i::CartesianIndex)
            return $fn($(args...), Tuple(i)...)
        end

        function $fn($(args...), i::Integer...)
            idcs = to_indices($(args[1]), i)
            S = IndexStyle($(args[1]))
            if isa(S, IndexLinear)
                return $fn($(args...), _to_linear_index($(args[1]), idcs...))
            else
                return $fn($(args...), _to_subscript_indices($(args[1]), idcs...))
            end
        end
    end
end
```
which works just fine (as julia code), but doesn't pass bootstrap. Any other ideas or should I just unroll those?

Reply:
you could probably fix the issue by moving all of the complicated ones out of genericmemory.jl. I'm not sure why it would fail bootstrap, but `genericmemory` is pretty early in it, so lots of stuff isn't defined. Specifically, I don't think `CarteisianIndex` exists yet...

Reply:
Thanks! no it didn't, and neither did iteration etc ;)

I added this code block in `base/multidimensional.jl` in the end.